Dockerfile

```diff
@@ -0,0 +1,138 @@
+# ARG base_image=ubuntu:18.04
+ARG base_image=python:3.8
+FROM $base_image
+
+USER root
+
+# Version strings and default paths
+ENV SPARK_VERSION=2.4.5 \
+    HADOOP_VERSION=2.7 \
+    HADOOP_HOME=/usr/local/hdp \
+    SPARK_HOME=/usr/local/spark \
+    DBT_HOME=/usr/local/dbt-spark \
+    HOME=/home
+WORKDIR $HOME
+
+### Use spark/conf instead of hadoop/shared/conf:
+ENV HADOOP_CONF_DIR /usr/local/spark/conf
+
+# Env tuning
+ENV PYTHONUNBUFFERED=1 \
+    PYTHONIOENCODING=utf-8 \
+    LANG=C.UTF-8
+
+# Install core libraries and dependencies
+RUN apt-get -y update && \
+    apt-get install -y \
+    apt-transport-https \
+    apt-utils \
+    ca-certificates \
+    ca-certificates-java \
+    iptables \
+    libcurl4-openssl-dev \
+    libsasl2-dev \
+    libssl-dev \
+    libxml2-dev \
+    lxc \
+    openssh-client \
+    net-tools \
+    wget && \
+    rm -rf /var/lib/apt/lists/*
+
+# Install Java
+RUN apt-get -y update && \
+    apt-get install -y \
+    default-jre && \
+    rm -rf /var/lib/apt/lists/* && \
+    echo JAVA_HOME=$JAVA_HOME && \
+    echo "which java=`which java`" && \
+    java -version
+
+# RUN update-alternatives --config java
+# RUN apt-get -y update && \
+#     apt-get install -y \
+#     -t jessie-backports \
+#     openjdk-8-jre-headless && \
+#     rm -rf /var/lib/apt/lists/*
+
+# Download spark
+RUN cd /tmp && \
+    SPARK_NAME=spark-${SPARK_VERSION}-bin-hadoop${HADOOP_VERSION} && \
+    curl -o $SPARK_NAME.tgz http://mirrors.sonic.net/apache/spark/spark-${SPARK_VERSION}/$SPARK_NAME.tgz && \
+    tar -xzf $SPARK_NAME.tgz -C /usr/local --owner root --group root --no-same-owner && \
+    rm $SPARK_NAME.tgz && \
+    cd /usr/local && \
+    ln -s $SPARK_NAME spark
+
+# Spark config
+ENV PYTHONPATH=$SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.10.7-src.zip \
+    SPARK_OPTS="--driver-java-options=-Xms1024M --driver-java-options=-Xmx4096M --driver-java-options=-Dlog4j.logLevel=info" \
+    PATH=$PATH:$SPARK_HOME/bin
+
+# Install mysql server and drivers (for hive metastore)
+RUN apt-get update && \
+    apt-get install -y \
+    default-libmysqlclient-dev \
+    default-mysql-server
+
+# Install Hadoop
+RUN cd /usr/local && \
+    HADOOP_MINOR_VERSION=2.7.7 && \
+    curl -o hadoop-$HADOOP_MINOR_VERSION.tar.gz https://archive.apache.org/dist/hadoop/common/hadoop-$HADOOP_MINOR_VERSION/hadoop-$HADOOP_MINOR_VERSION.tar.gz && \
+    tar -xzf hadoop-$HADOOP_MINOR_VERSION.tar.gz && \
+    mv hadoop-$HADOOP_MINOR_VERSION $HADOOP_HOME && \
+    chown -R root:root $HADOOP_HOME && \
+    chmod -R 777 $HADOOP_HOME && \
+    rm hadoop-$HADOOP_MINOR_VERSION.tar.gz
+
+# Copy hadoop libraries for AWS & S3 to spark classpath
```

Comment:
should we make that optional ? maybe we could use a Dockerfile.template ?

Reply:
I'm not familiar with `Dockerfile.template`. I do think that both AWS and Azure libraries (jars) are needed in the 'base/core' image. They enable core functionality that otherwise would be a blocker for even basic usage and testing.

One simplification I could make here, **if** we didn't want the rest of the hadoop binaries, is that I could use a multi-image build, downloading hadoop binaries in a ephemeral image and then only copy the files we need into the spark classpath.

Thoughts?

Reply:
I just imagine AWS users might not need Azure / GCP jars and vice versa.
Or might not need any if they are just using the local filesystem / hdfs ?

I have just seen in past projects, people using tokens inside their Dockerfile that they replace using sed to customize a "base" dockerfile. Could also have a base docker image and then an aws, azure and gcp image built from it.

All that said, this PR should probably just focus on providing a local environment (through Docker) that is good enough for (integration) testing, we could look into making the image smaller later.

Anyways! I'm going to play around with this image over this week end and give you more feedback :)